### PR TITLE
[Main Course] - Tensorflow Recommender

### PR DESCRIPTION
**Pull request by Tina0330**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] This change requires a documentation update

# How Has This Been Tested?

- [x] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: docs/Main_Course/Deep_Learning/2021-12-10-Tensorflow-Recommender.md [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+
+## 3 源码解析
+
+现实推荐系统分为两大部分：
+
+1. 从大量的潜在推荐条目中选择可能性比较大的items，这一过程叫做信息提取(retrieval)。
+
+2. 对于提取模型的结果，我们还需要排序来缩小选择最有可能被用户选择的items。
+
+这一部分，我们先介绍第一阶段的信息提取模型。信息提取模型又包含两个子模型，查询模型和候选模型，对应上述的例子就是用户矩阵和候选条目矩阵，通过计算两个子模型的乘积，得到的`query-candidate affinity score`就是反应查询和候选条目之间的匹配程度，即用户喜欢推荐条目的可能性。
```
Comment (EchoTang04) [inline docs/Main_Course/Deep_Learning/2021-12-10-Tensorflow-Recommender.md:61]: “查询模型和候选模型，对应上述的例子就是用户矩阵和候选条目矩阵”，这两个矩阵必须维度相等吧？所以当我们用n维向量表示每一个用户的喜好的时候，必须也用n维的向量来衡量电影类别是嘛？

File: docs/Main_Course/Deep_Learning/2021-12-10-Tensorflow-Recommender.md [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+
```
Comment (EchoTang04) [inline docs/Main_Course/Deep_Learning/2021-12-10-Tensorflow-Recommender.md:62]: 实际使用的时候怎么考虑维度n的选取呢？

**Review by EchoTang04 — APPROVED** [review]
LGTM, 感谢分享~ 留了几个小问题，可以讨论一下

File: docs/Main_Course/Deep_Learning/2021-12-10-Tensorflow-Recommender.md (hunk 1 above)
Comment (Tina0330) [reply docs/Main_Course/Deep_Learning/2021-12-10-Tensorflow-Recommender.md:61]: 在这个retrievel理论中是的，每个embedding维度在两个模型中是要有对应的设置的，否则无法计算affinity score

Comment (EchoTang04) [reply docs/Main_Course/Deep_Learning/2021-12-10-Tensorflow-Recommender.md:61]: 这个还挺有意思的。感觉两个模型角度很不一样，但是必须embed维度相等。还是那个问题，怎么判断n是多少最合适呢？

File: docs/Main_Course/Deep_Learning/2021-12-10-Tensorflow-Recommender.md (hunk 2 above)
Comment (Tina0330) [reply docs/Main_Course/Deep_Learning/2021-12-10-Tensorflow-Recommender.md:62]: 这个取决于文本数据和词数据的大小的。有很多种embedding的方式，选的标准可能不太一样，这我也不太确定，需要研究学习一下 ：）

File: docs/Main_Course/Deep_Learning/2021-12-10-Tensorflow-Recommender.md [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+---
+
+# TensorFlow推荐系统
+
+!!! info
+    作者：Tina，发布于2021-12-10，阅读时间：约6分钟，微信公众号文章链接：[:fontawesome-solid-link:]()
+
+## 1 前言
+
+当我们浏览在各个平台时会发现**为你推荐**功能。比如Youtube推荐爱看的视频，音乐软件为你提供你可能喜欢的音乐等。其实这一功能的背后涉及的原理就是人工智能的推荐系统。今天我们将介绍Tensorflow推荐系统模型的库——[Tensoflow Recommenders(TFRS)](https://www.tensorflow.org/recommenders?hl=en 'TensorFlow Recommenders')。
```
Comment (jhqian0909) [inline docs/Main_Course/Deep_Learning/2021-12-10-Tensorflow-Recommender.md:12]: 可以去掉在。加粗blog可能有显示问题。

File: docs/Main_Course/Deep_Learning/2021-12-10-Tensorflow-Recommender.md [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+- [Keras各种Callbacks介绍](https://mp.weixin.qq.com/s/N3_1lJ_a1gAVLyqsXUwN9Q)
+- [使用tf.keras自定义模型](https://mp.weixin.qq.com/s/z2uBxwe8UNDXWMDNS_k-Gg)
+- [在Apple Silicon Macs上安装TensorFlow](https://mp.weixin.qq.com/s/QHZW_lcMo_nPQWDYPDGREg)
+- [读《30天吃掉那只TensorFlow2》](https://mp.weixin.qq.com/s/cw2DW7al5nJV93roAN_gwg)
+
+## 2 推荐原理
+
+这里我们用电影推荐的例子来讲解推荐系统的原理。
+
+对于现有的四个用户和五部类型不同的电影，首先，我们需要对创建用户画像和电影类别的定义，这一步是为了区分数据，将现实特征转化为可计算的变量。对于现有的用户数据的电影数据，我们如何给用户D推荐她可能喜欢的电影呢？
```
Comment (jhqian0909) [inline docs/Main_Course/Deep_Learning/2021-12-10-Tensorflow-Recommender.md:25]: 去掉对

Comment (jhqian0909) [reply docs/Main_Course/Deep_Learning/2021-12-10-Tensorflow-Recommender.md:25]: 还有一些typo

File: docs/Main_Course/Deep_Learning/2021-12-10-Tensorflow-Recommender.md [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+训练并评估模型结果：
+
+```Python
+model = MovielensModel(user_model, movie_model)
+model.compile(optimizer=tf.keras.optimizers.Adagrad(learning_rate=0.1))
+#shuffle data and get samples
+cached_train = train.shuffle(100_000).batch(8192).cache()
+cached_test = test.batch(4096).cache()
+
+model.fit(cached_train, epochs=3
```
Comment (jhqian0909) [inline docs/Main_Course/Deep_Learning/2021-12-10-Tensorflow-Recommender.md:192]: ）

File: docs/Main_Course/Deep_Learning/2021-12-10-Tensorflow-Recommender.md [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+    "movie_title": x["movie_title"],
+    "user_id": x["user_id"],
+})
+movies = movies.map(lambda x: x["movie_title"])
+```
+设定训练集和测试数据集：
+
+```Python
+tf.random.set_seed(42)
+shuffled = ratings.shuffle(100_000, seed=42, reshuffle_each_iteration=False)
```
Comment (jhqian0909) [inline docs/Main_Course/Deep_Learning/2021-12-10-Tensorflow-Recommender.md:109]: 多了下划线？

Comment (Tina0330) [reply docs/Main_Course/Deep_Learning/2021-12-10-Tensorflow-Recommender.md:109]: 特意看了一下，这个就是100，000的意思，可能是为了更好的辨识吧

File: docs/Main_Course/Deep_Learning/2021-12-10-Tensorflow-Recommender.md [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+</figure>
+
+
+## 3 源码解析
+
+现实推荐系统分为两大部分：
+
+1. 从大量的潜在推荐条目中选择可能性比较大的items，这一过程叫做信息提取(retrieval)。
+
+2. 对于提取模型的结果，我们还需要排序来缩小选择最有可能被用户选择的items。
```
Comment (BulletTech2021) [inline docs/Main_Course/Deep_Learning/2021-12-10-Tensorflow-Recommender.md:59]: 这一过程叫rank

File: docs/Main_Course/Deep_Learning/2021-12-10-Tensorflow-Recommender.md [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+<figure>
+  <img src="https://cdn.jsdelivr.net/gh/BulletTech2021/Pics/img/function_tfrs.png" width="500" />
+</figure>
+
+
+## 3 源码解析
+
+现实推荐系统分为两大部分：
+
+1. 从大量的潜在推荐条目中选择可能性比较大的items，这一过程叫做信息提取(retrieval)。
```
Comment (BulletTech2021) [inline docs/Main_Course/Deep_Learning/2021-12-10-Tensorflow-Recommender.md:57]: 注意中英文标点符号，括号用中文的

File: docs/Main_Course/Deep_Learning/2021-12-10-Tensorflow-Recommender.md [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+_, titles = index(tf.constant(["9"]))
+print(f"Recommendations for user 9: {titles[0, :3]}")
+```
+<figure>
+  <img src="https://cdn.jsdelivr.net/gh/BulletTech2021/Pics/img/recom.png" width="500" />
+</figure>
+
+## 4 总结
+
+TensorFlow 为NLP的学习提供了非常丰富且强大的资源，感兴趣的朋友可以将这些模型运用到现有的文本数据中，去探究一些有趣的惊喜吧！
```
Comment (BulletTech2021) [inline docs/Main_Course/Deep_Learning/2021-12-10-Tensorflow-Recommender.md:223]: 这不是NLP

File: docs/Main_Course/Deep_Learning/2021-12-10-Tensorflow-Recommender.md (hunk 7 above)
Comment (Tina0330) [reply docs/Main_Course/Deep_Learning/2021-12-10-Tensorflow-Recommender.md]: 已修改

File: docs/Main_Course/Deep_Learning/2021-12-10-Tensorflow-Recommender.md (hunk 5 above)
Comment (Tina0330) [reply docs/Main_Course/Deep_Learning/2021-12-10-Tensorflow-Recommender.md]: 已修改
